simd/unit_tests/include/TestSIMD_WhereExpressions.hpp

```diff
@@ -0,0 +1,195 @@
+    mask_type mask(true);
+    mask[idx] = false;
+
+    simd_type dst;
+    index_type index;
+    simd_type expected_result;
+    for (std::size_t i = 0; i < nlanes; ++i) {
+      dst[i]             = (2 + (i * 2));
+      index[i]           = i;
+      expected_result[i] = (mask[i]) ? src[index[i]] : (2 + (i * 2));
+    }
+    where(mask, dst).gather_from(src, index);
+
+    host_check_equality(expected_result, dst, nlanes);
+  }
+}
+
+template <class Abi, typename DataType>
+inline void host_check_where_expr() {
+  host_check_where_expr_scatter_to<Abi, DataType>();
+  host_check_where_expr_gather_from<Abi, DataType>();
+}
+
+template <typename Abi, typename... DataTypes>
+inline void host_check_where_expr_all_types(
+    Kokkos::Experimental::Impl::data_types<DataTypes...>) {
+  (host_check_where_expr<Abi, DataTypes>(), ...);
+}
+
+template <typename... Abis>
+inline void host_check_where_expr_all_abis(
+    Kokkos::Experimental::Impl::abi_set<Abis...>) {
+  using DataTypes = Kokkos::Experimental::Impl::data_type_set;
+  (host_check_where_expr_all_types<Abis>(DataTypes()), ...);
+}
+
+template <typename Abi, typename DataType>
+KOKKOS_INLINE_FUNCTION void device_check_where_expr_scatter_to() {
+  using simd_type  = Kokkos::Experimental::simd<DataType, Abi>;
+  using index_type = Kokkos::Experimental::simd<std::int32_t, Abi>;
+  using mask_type  = typename simd_type::mask_type;
+
+  std::size_t nlanes = simd_type::size();
+  DataType init[]    = {11, 13, 17, 19, 23, 29, 31, 37};
+  simd_type src;
+  src.copy_from(init, Kokkos::Experimental::element_aligned_tag());
+
+  for (std::size_t idx = 0; idx < nlanes; ++idx) {
+    mask_type mask(true);
+    mask[idx] = false;
+
+    DataType dst[8] = {0};
+    index_type index;
+    simd_type expected_result;
+    for (std::size_t i = 0; i < nlanes; ++i) {
+      dst[i]             = (2 + (i * 2));
+      index[i]           = i;
+      expected_result[i] = (mask[i]) ? src[index[i]] : (2 + (i * 2));
+    }
+    where(mask, src).scatter_to(dst, index);
+
+    simd_type dst_simd;
+    dst_simd.copy_from(dst, Kokkos::Experimental::element_aligned_tag());
+
+    device_check_equality(expected_result, dst_simd, nlanes);
+  }
+}
+
+template <typename Abi, typename DataType>
+KOKKOS_INLINE_FUNCTION void device_check_where_expr_gather_from() {
+  using simd_type  = Kokkos::Experimental::simd<DataType, Abi>;
+  using index_type = Kokkos::Experimental::simd<std::int32_t, Abi>;
+  using mask_type  = typename simd_type::mask_type;
+
+  std::size_t nlanes = simd_type::size();
+  DataType src[]     = {11, 13, 17, 19, 23, 29, 31, 37};
+
+  for (std::size_t idx = 0; idx < nlanes; ++idx) {
+    mask_type mask(true);
+    mask[idx] = false;
+
+    simd_type dst;
+    index_type index;
+    simd_type expected_result;
+    for (std::size_t i = 0; i < nlanes; ++i) {
+      dst[i]             = (2 + (i * 2));
+      index[i]           = i;
+      expected_result[i] = (mask[i]) ? src[index[i]] : (2 + (i * 2));
```

Comment:
```suggestion
      expected_result[i] = (mask[i]) ? src[index[i]] : dst[i];
```

Reply:
Same comment everywhere

Reply:
Changed in [09fd0c2](https://github.com/kokkos/kokkos/pull/6220/commits/09fd0c20743be077e061510fe7a53321675452e1).